Version 0.1

mecha opened the pull request:
Initial release.

Contains solutions for #2 and #3.

mecha committed:
Namespace prompt is more convenient

Now expects single slash and doubles them after prompt.
Also adds a trailing slash.

mecha committed:
Updated readme file

mecha committed:
Added changelog template file

mecha committed:
Updated changelog

XedinUnknown reviewed (commented):
How about adding a Composer `build` script? This would save typing the path to `vendor/bin/phing`.

XedinUnknown commented:
After creation finished, there was no `master` branch. I would much prefer there to be both the `develop` and `master`, and to be on `master` after creation.

XedinUnknown commented:
I think `build.xml` needs to be removed after creation, too.

mecha commented:
@XedinUnknown Unfortunately, due to a [Composer bug](https://github.com/composer/composer/issues/3299) the build script would not stop at the prompts.

There should be both a `master` and `develop` branch - but even so I agree to be on `master` by default. I will confirm the absence of `master` and make it default if not already so.

XedinUnknown commented:
> Unfortunately, due to a Composer bug the build script would not stop at the prompts.

@mecha, fair enough. Maybe add this in a "Known Issues" section? Or, just create an issue.

XedinUnknown commented:
@mecha, when you fix the absence of `master`, and make it the default, as well as remove the `build.xml` file, maybe also add all files to index?

mecha commented:
@XedinUnknown That's a good idea. Maybe even make an "Initial commit"?

XedinUnknown commented:
Maybe the commit is too much, since we may wanna add some other config files.

XedinUnknown commented:
Single slashes didn't get converted to double slashes for me, and also the slashes weren't added at the end. I entered `Dhii\Wp\I18n`.

mecha committed:
Fixed some formatting

XedinUnknown commented:
Another thing that could be helpful:
`$ git remote add origin https://github.com/dhii/new-project`

mecha committed:
Renamed some filesets for added clarity

mecha committed:
Fixed master branch not created and not default

mecha committed:
Added build file self-destruction

mecha committed:
Renamed tokens in netbeans project.properties file

XedinUnknown commented:
@mecha, looks like the composer deps didn't get installed at project creation time. That's why it's so fast.

mecha committed:
Fixed slash replacement for namespaces

mecha committed:
Composer updates dependencies during finalization

mecha committed:
All files are added to the Git index

mecha committed:
Composer lock file is deleted before updating deps

mecha committed:
Removed unecessary fileset

mecha committed:
Build script can now add a Git remote

mecha committed:
Composer update command uses passthru

mecha commented:
@XedinUnknown Test now. All issues should be addressed.

mecha commented:
An FYI to all testers, the command for testing is:

```
composer create-project dhii/project-bootstrap:dev-release/0.1 my-project
```

After releasing, the version constraint can be eliminated.

mecha committed:
Fixed selfdestruct target not being invoked

mecha committed:
Add known issues to readme

mecha committed:
Fixed Git branch creation

mecha committed:
Moved default license file to `licenses` dir

mecha committed:
Added GPL-3.0 license file

mecha committed:
Added build functions file

mecha committed:
Build process now templates correct license file

mecha committed:
Add newline at end of test bootstrap file

Fixes #3

mecha commented:
@XedinUnknown Issue #3 has been addressed and #2 has been implemented in the build process.

mecha committed:
Build process uses intelligent guessing for vendor ...

and package names via the author and project names.

mecha committed:
Readme file is templated with project and author names

mecha committed:
Documented `toCamelCase` build function

mecha committed:
Namespace generation function now uses vendor

mecha committed:
Organization name now generated from vendor

mecha committed:
Renamed test suites

mecha committed:
Added scripts to `composer.json` template

XedinUnknown commented:
Working very well. No more suggestions or required questions. I think this is ready for merge. Will just try it on a couple of other cases, just to make sure.

XedinUnknown commented:
Add `composer phing` ability with `scripts`, if possible, please, @mecha.

mecha committed:
Amended message shown after project creation

mecha commented:
@XedinUnknown Not possible. See [this comment](https://github.com/Dhii/bootstrap/pull/1#issuecomment-286836923) further up.

What I _will_ do is add test templates in the test sub-directories, as agreed in an earlier conversation.

XedinUnknown commented:
When I write "Dhii" for vendor, the human readables e.g. README and License should show it as written, i.e. "Dhii", but Composer, NB proj, and such, should see the slug, i.e. "dhii". Right now, it didn't happen, and NB project name, as well as the Composer package name, had "Dhii", i.e. verbatim.

mecha committed:
Added test templates

mecha commented:
Ah I see. You entered "Dhii" for the _vendor_ name. Right now, the script expects you to enter the vendor name correctly and does not do any manipulations to it. What you want to do is enter "Dhii" as the _author_.

XedinUnknown commented:
@mecha, yes, I enter "Dhii" or "Dhii Team" or whatever I want the author to be as the author, and that goes correctly. The vendor name used in configuration etc, versus what is used for humans, must be sanitized, e.g. lower- and kebab-cased. The default of that should be derived from the author.

XedinUnknown commented:
The dummy test's `createInstance()` method mocks `methods()` method. I think this should not be there.

XedinUnknown commented:
Also, the FQNs used in dummy tests contain stuff like `_vendor\_package\FuncTest\_ns`. Don't we already know the root namespace to substitute `_vendor\_package` with? The `_ns` looks extra here, as the first classes are likely to be right in the root NS + test NS.

XedinUnknown commented:
Another useful feature could be to collapse dashes when converting HR -> slug. Example: "Data - Tree - Abstract" becomes `data-tree-abstract` instead of `data---tree---abstract`. Should be very easy to achieve: convert any number of symbols from set, such as `[ _-+\/]`, into one symbol.

XedinUnknown commented:
Using the [`prefer-stable`] flag could really improve bootstrap speed, as it would cause cached versions to be used for many packages.

[`prefer-stable`]:                          https://getcomposer.org/doc/04-schema.md#prefer-stable

XedinUnknown committed:
Now preferring stable packages

mecha committed:
Improved kebab case conversion with regex.

All consequent invalid characters are now collapsed into a single dash.

mecha committed:
Vendor name is now kebab cased.

mecha committed:
Removed the func test mock `methods()` method

mecha committed:
Test files are now templated with the namespace

mecha commented:
@XedinUnknown Addressed the issues.

XedinUnknown commented on build-functions.php:
```diff
@@ -29,7 +29,7 @@ function kebabCaseToUcWords($kebab)
 function toKebabCase($string)
 {
-    return implode('-', explode(' ', strtolower($string)));
+    return preg_replace('/[ _\\-\\+\\/\\\\]+/', '-', strtolower($string));
```
This is wrong. See [example](http://www.phpliveregex.com/p/kCA). You should probably first replace all separation chars, e.g. `- +_/\\`, with `-`, and then collapse all collapsible chars similarly to what you did, but capture the actual char, and replace with a single instance of that char, e.g. `$0`.

mecha replied:
Didn't understand the second part, after "and then collapse ...".

But note that the example on the site needs to be de-escaped. [See now](http://www.phpliveregex.com/p/kCB).

mecha replied:
Amended the function. Now uses a `collapseChars()` function.

mecha committed:
Project and package name detected from dir name

mecha committed:
Fixed use of vendor, author and organization.

- Organization is now prompted for as "Organization name"
- Vendor is now prompted for as "Package Vendor"

mecha committed:
Organization is now deduced from author name

XedinUnknown committed:
Merge pull request #5 from Dhii/task/prefer-stable

Now preferring stable packages

XedinUnknown commented:
@mecha, I think it may also be worth adding `prefer-stable` to this repo's Composer config as well.

mecha commented:
@XedinUnknown Agree.

mecha commented:
@XedinUnknown Perhaps also use `master` for `dhii/php-cs-fixer-config`?

XedinUnknown commented:
I don't think ready yet. We'd have to bump up the required PHP version for that. And we still need PHP 5.3 for many of the interfaces.

mecha committed:
Added `prefer-stable` to composer.json template

mecha committed:
Merge branch 'release/0.1' of https://github.com/Dhii/bootstrap into release/0.1

mecha commented:
@XedinUnknown Added the `prefer-stable` option in the composer config template.

Any last changes? I'd like to merge this and release.

XedinUnknown commented:
OMG this merge was a mistake. @mecha, I wrote to add
> `prefer-stable` to *this* repo's Composer config as well

I have no idea how you added it to the template, especially since I already did that earlier, in https://github.com/Dhii/bootstrap/commit/13c451de75227efea4e1208b079504ecd106e1e8.

mecha commented:
No, _I_ added it to the `composer.json` template. You added it to the `composer.json` of the bootstrap.

XedinUnknown commented:
You didn't click the commit link?
